tests/linearizability/client.go

```diff
@@ -76,16 +84,45 @@ func (c *recordingClient) Put(ctx context.Context, key, value string) error {
 	callTime := time.Now()
 	resp, err := c.client.Put(ctx, key, value)
 	returnTime := time.Now()
+	if err != nil {
+		c.failed = append(c.failed, porcupine.Operation{
+			ClientId: int(c.id),
+			Input:    EtcdRequest{Op: Put, Key: key, PutData: value},
+			Call:     callTime.UnixNano(),
+			Output:   EtcdResponse{Err: err},
+		})
+		c.id = clientId.Add(1) - 1
+		return err
+	}
 	var revision int64
 	if resp != nil && resp.Header != nil {
 		revision = resp.Header.Revision
 	}
 	c.operations = append(c.operations, porcupine.Operation{
-		ClientId: c.id,
-		Input:    etcdRequest{op: Put, key: key, putData: value},
+		ClientId: int(c.id),
+		Input:    EtcdRequest{Op: Put, Key: key, PutData: value},
 		Call:     callTime.UnixNano(),
-		Output:   etcdResponse{err: err, revision: revision},
+		Output:   EtcdResponse{Err: err, Revision: revision},
 		Return:   returnTime.UnixNano(),
 	})
 	return nil
 }
+
+func (c *recordingClient) Operations() []porcupine.Operation {
+	operations := make([]porcupine.Operation, 0, len(c.operations)+len(c.failed))
+	var maxTime int64
+	for _, op := range c.operations {
+		operations = append(operations, op)
+		if op.Return > maxTime {
+			maxTime = op.Return
+		}
+	}
+	for _, op := range c.failed {
+		if op.Call > maxTime {
```

Comment:
can this really happen? not sure if it would be better to panic here when the clocks tick the other way and we have Returns finishing before Calls.